# Table of Content
1. Importing libraries and dataset
2. Understanding data
3. Data quality checks
4. Data Cleaning after data quality checks¶
5. Data sense checks
6. Data Cleaning after data sense checks
7. Exporting cleaned dataframe¶

# 1. Importing libraries and dataset

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
# Defining path for data import
path = r'/Users/frederikeschulz-mullensiefen/Desktop/Masterfolder_Hotel Bookings'

In [4]:
# Define needed columns of dataframe
bookings_list = ['hotel', 'is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'meal', 'country', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'deposit_type', 'days_in_waiting_list', 'customer_type', 'adr', 'reservation_status']

In [5]:
# Import hotel bookings dataset 
df_bookings = pd.read_csv(os.path.join(path, '02_Data', 'Original Data', 'hotel_bookings.csv'), usecols = bookings_list)

# 2. Understanding data

In [6]:
# Checking dataframe shape
df_bookings.shape

(119390, 22)

In [7]:
# Setting no limit on max columns to display
pd.set_option('display.max_columns', None)

In [8]:
# Checking first rows of dataframe
df_bookings.head(10)

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   
5  Resort Hotel            0         14               2015               July   
6  Resort Hotel            0          0               2015               July   
7  Resort Hotel            0          9               2015               July   
8  Resort Hotel            1         85               2015               July   
9  Resort Hotel            1         75               2015               July   

   arrival_date_week_number  stays_in_weekend_nights  stays_in_week_nights  \
0                        27                        0                     0   
1                        27                        0                     0   
2                        27                        0                     1   
3                        27                        0                     1   
4                        27                        0                     2   
5                        27                        0                     2   
6                        27                        0                     2   
7                        27                        0                     2   
8                        27                        0                     3   
9                        27                        0                     3   

   adults  children  babies meal country  is_repeated_guest  \
0       2       0.0       0   BB     PRT                  0   
1       2       0.0       0   BB     PRT                  0   
2       1       0.0       0   BB     GBR                  0   
3       1       0.0       0   BB     GBR                  0   
4       2       0.0       0   BB     GBR                  0   
5       2       0.0       0   BB     GBR                  0   
6       2       0.0       0   BB     PRT                  0   
7       2       0.0       0   FB     PRT                  0   
8       2       0.0       0   BB     PRT                  0   
9       2       0.0       0   HB     PRT                  0   

   previous_cancellations  previous_bookings_not_canceled  booking_changes  \
0                       0                               0                3   
1                       0                               0                4   
2                       0                               0                0   
3                       0                               0                0   
4                       0                               0                0   
5                       0                               0                0   
6                       0                               0                0   
7                       0                               0                0   
8                       0                               0                0   
9                       0                               0                0   

  deposit_type  days_in_waiting_list customer_type    adr reservation_status  
0   No Deposit                     0     Transient    0.0          Check-Out  
1   No Deposit                     0     Transient    0.0          Check-Out  
2   No Deposit                     0     Transient   75.0          Check-Out  
3   No Deposit                     0     Transient   75.0          Check-Out  
4   No Deposit                     0     Transient   98.0          Check-Out  
5   No Deposit                     0     Transient   98.0          Check-Out  
6   No Deposit                     0     Transient  107.0          Check-Out  
7   No Deposit                     0     Transient  103.0          Check-O

# 3. Data quality checks

In [9]:
# Checking data types 
df_bookings.dtypes

hotel                              object
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
booking_changes                     int64
deposit_type                       object
days_in_waiting_list                int64
customer_type                      object
adr                               float64
reservation_status                 object
dtype: object

In [10]:
# Finding missing data 
df_bookings.isnull().sum()

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            4
babies                              0
meal                                0
country                           488
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
booking_changes                     0
deposit_type                        0
days_in_waiting_list                0
customer_type                       0
adr                                 0
reservation_status                  0
dtype: int64

In [10]:
# Finding duplicates
df_dups = df_bookings[df_bookings.duplicated()]

In [11]:
# Showing duplicates
df_dups

hotel  is_canceled  lead_time  arrival_date_year  \
5       Resort Hotel            0         14               2015   
22      Resort Hotel            0         72               2015   
43      Resort Hotel            0         70               2015   
129     Resort Hotel            0        100               2015   
130     Resort Hotel            0        100               2015   
...              ...          ...        ...                ...   
119349    City Hotel            0        186               2017   
119352    City Hotel            0         63               2017   
119353    City Hotel            0         63               2017   
119354    City Hotel            0         63               2017   
119373    City Hotel            0        175               2017   

       arrival_date_month  arrival_date_week_number  stays_in_weekend_nights  \
5                    July                        27                        0   
22                   July                        27                        2   
43                   July                        27                        2   
129                  July                        27                        1   
130                  July                        27                        1   
...                   ...                       ...                      ...   
119349             August                        35                        0   
119352             August                        35                        0   
119353             August                        35                        0   
119354             August                        35                        0   
119373             August                        35                        1   

        stays_in_week_nights  adults  children  babies meal country  \
5                          2       2       0.0       0   BB     GBR   
22                         4       2       0.0       0   BB     PRT   
43                         3       2       0.0       0   HB     ROU   
129                        1       2       0.0       0   BB     FRA   
130                        1       2       0.0       0   BB     FRA   
...                      ...     ...       ...     ...  ...     ...   
119349                     3       2       0.0       0   BB     DEU   
119352                     3       3       0.0       0   BB     SWE   
119353                     3       3       0.0       0   BB     SWE   
119354                     3       3       0.0       0   BB     SWE   
119373                     3       1       0.0       0   BB     NLD   

        is_repeated_guest  previous_cancellations  \
5                       0                       0   
22                      0                       0   
43                      0                       0   
129                     0                       0   
130                     0                       0   
...                   ...                     ...   
119349                  0                       0   
119352                  0                       0   
119353                  0                       0   
119354                  0                       0   
119373                  0                       0   

        previous_bookings_not_canceled  booking_changes deposit_type  \
5                                    0                0   No Deposit   
22                                   0                1   No Deposit   
43                                   0                0   No Deposit   
129                                  0                0   No Deposit   
130                                  0                0   No Deposit   
...                                ...              ...          ...   
119349                               0                0   No Deposit   
119352                               0                0   No Deposit   
119353                               0                0   No Deposit   
119354                               0 

In [12]:
# Checking for mixed-type data
for col in df_bookings.columns.tolist():
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_bookings[weird]) > 0:
    print (col)

/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap

country


/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/3279077479.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)


# 4. Data Cleaning after data quality checks

In [13]:
# Correcting datatypes
df_bookings['is_canceled'] = df_bookings['is_canceled'].astype('str')
df_bookings['is_repeated_guest'] = df_bookings['is_repeated_guest'].astype('str')
df_bookings['country'] = df_bookings['country'].astype('str')

In [14]:
# Checking if data type change was sucessful
df_bookings.dtypes

hotel                              object
is_canceled                        object
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
is_repeated_guest                  object
previous_cancellations              int64
previous_bookings_not_canceled      int64
booking_changes                     int64
deposit_type                       object
days_in_waiting_list                int64
customer_type                      object
adr                               float64
reservation_status                 object
dtype: object

In [15]:
# Checking if mixed-type data are gone
for col in df_bookings.columns.tolist():
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_bookings[weird]) > 0:
    print (col)

/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/1734789829.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/1734789829.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/1734789829.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap(type) != df_bookings[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/1734789829.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_bookings[[col]].applymap

In [16]:
# Removing missing values in country column
df_bookings_clean1 = df_bookings[df_bookings['country'].isnull() == False]

In [17]:
df_bookings_clean1.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          4
babies                            0
meal                              0
country                           0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
reservation_status                0
dtype: int64

In [18]:
# Removing missing values in children column
df_bookings_clean2 = df_bookings_clean1[df_bookings_clean1['children'].isnull() == False]

In [19]:
# Checking successul removal of null values
df_bookings_clean2.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
reservation_status                0
dtype: int64

# 5. Data sense checks

In [20]:
# Checking descriptive statistics
df_bookings_clean2.describe()

lead_time  arrival_date_year  arrival_date_week_number  \
count  119386.000000      119386.000000             119386.000000   
mean      104.014801        2016.156593                 27.165003   
std       106.863286           0.707456                 13.605334   
min         0.000000        2015.000000                  1.000000   
25%        18.000000        2016.000000                 16.000000   
50%        69.000000        2016.000000                 28.000000   
75%       160.000000        2017.000000                 38.000000   
max       737.000000        2017.000000                 53.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119386.000000         119386.000000  119386.000000   
mean                  0.927605              2.500310       1.856390   
std                   0.998618              1.908289       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  previous_cancellations  \
count  119386.000000  119386.000000           119386.000000   
mean        0.103890       0.007949                0.087121   
std         0.398561       0.097438                0.844350   
min         0.000000       0.000000                0.000000   
25%         0.000000       0.000000                0.000000   
50%         0.000000       0.000000                0.000000   
75%         0.000000       0.000000                0.000000   
max        10.000000      10.000000               26.000000   

       previous_bookings_not_canceled  booking_changes  days_in_waiting_list  \
count                   119386.000000    119386.000000         119386.000000   
mean                         0.137102         0.221131              2.321227   
std                          1.497462         0.652315             17.595011   
min                          0.000000         0.000000              0.000000   
25%                          0.000000         0.000000              0.000000   
50%                          0.000000         0.000000              0.000000   
75%                          0.000000         0.000000              0.000000   
max                         72.000000        21.000000            391.000000   

                 adr  
count  119386.000000  
mean      101.833541  
std        50.534664  
min        -6.380000  
25%        69.290000  
50%        94.590000  
75%       126.000000  
max      5400.000000

From the descriptive statistics, almost all columns seem to make sense. There are a few things that would need further investigation:

-55 adults for one booking seems excessive (though not impossible)

-391 days in the waiting list seems rather long (though not impossible)

-An average daily rate of below zero does not make sense

-It should be checked whether the total number of nights of stay is above zero

-It should be checked whether the total number of guests is above zero

In [21]:
# Further sense check for adr below zero
sense_check_adr = df_bookings_clean2[df_bookings_clean2['adr']<0]

In [22]:
# Checking count of observations where adr < 0
sense_check_adr

hotel is_canceled  lead_time  arrival_date_year  \
14969  Resort Hotel           0        195               2017   

      arrival_date_month  arrival_date_week_number  stays_in_weekend_nights  \
14969              March                        10                        4   

       stays_in_week_nights  adults  children  babies meal country  \
14969                     6       2       0.0       0   BB     GBR   

      is_repeated_guest  previous_cancellations  \
14969                 1                       0   

       previous_bookings_not_canceled  booking_changes deposit_type  \
14969                               2                2   No Deposit   

       days_in_waiting_list    customer_type   adr reservation_status  
14969                     0  Transient-Party -6.38          Check-Out

As there is only one record with a negative adr, this record should be removed.

In [23]:
# Deriving new variable "total_guests"
df_bookings_clean2['total_guests'] = (df_bookings_clean2['adults'] + df_bookings_clean2['children'] + df_bookings_clean2['babies'])

/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/2332852051.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_clean2['total_guests'] = (df_bookings_clean2['adults'] + df_bookings_clean2['children'] + df_bookings_clean2['babies'])


In [24]:
# Deriving new variable "total_number_nights"
df_bookings_clean2['total_number_nights'] = (df_bookings_clean2['stays_in_weekend_nights'] + df_bookings_clean2['stays_in_week_nights'])

/var/folders/6r/_pj4fghj6jx_vml8b8yvngzh0000gn/T/ipykernel_13288/1809588303.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_bookings_clean2['total_number_nights'] = (df_bookings_clean2['stays_in_weekend_nights'] + df_bookings_clean2['stays_in_week_nights'])


In [25]:
# Check if derivation of new variables worked
df_bookings_clean2.head(25)

hotel is_canceled  lead_time  arrival_date_year arrival_date_month  \
0   Resort Hotel           0        342               2015               July   
1   Resort Hotel           0        737               2015               July   
2   Resort Hotel           0          7               2015               July   
3   Resort Hotel           0         13               2015               July   
4   Resort Hotel           0         14               2015               July   
5   Resort Hotel           0         14               2015               July   
6   Resort Hotel           0          0               2015               July   
7   Resort Hotel           0          9               2015               July   
8   Resort Hotel           1         85               2015               July   
9   Resort Hotel           1         75               2015               July   
10  Resort Hotel           1         23               2015               July   
11  Resort Hotel           0         35               2015               July   
12  Resort Hotel           0         68               2015               July   
13  Resort Hotel           0         18               2015               July   
14  Resort Hotel           0         37               2015               July   
15  Resort Hotel           0         68               2015               July   
16  Resort Hotel           0         37               2015               July   
17  Resort Hotel           0         12               2015               July   
18  Resort Hotel           0          0               2015               July   
19  Resort Hotel           0          7               2015               July   
20  Resort Hotel           0         37               2015               July   
21  Resort Hotel           0         72               2015               July   
22  Resort Hotel           0         72               2015               July   
23  Resort Hotel           0         72               2015               July   
24  Resort Hotel           0        127               2015               July   

    arrival_date_week_number  stays_in_weekend_nights  stays_in_week_nights  \
0                         27                        0                     0   
1                         27                        0                     0   
2                         27                        0                     1   
3                         27                        0                     1   
4                         27                        0                     2   
5                         27                        0                     2   
6                         27                        0                     2   
7                         27                        0                     2   
8                         27                        0                     3   
9                         27                        0                     3   
10                        27                        0                     4   
11                        27                        0                     4   
12                        27                        0                     4   
13                        27                        0                     4   
14                        27                        0                     4   
15                        27                        0                     4   
16                        27                        0                     4   
17                        27                        0                     1   
18                        27                        0                     1   
19                        27                        0                     4   
20                        27                        1                     4   
21                        27                        2                     4   
22                        27                        2                     4   
23      

In [26]:
# Sense check for observations with zero total guests
sense_check_guests = df_bookings_clean2[df_bookings_clean2['total_guests']<1]

In [27]:
sense_check_guests

hotel is_canceled  lead_time  arrival_date_year  \
2224    Resort Hotel           0          1               2015   
2409    Resort Hotel           0          0               2015   
3181    Resort Hotel           0         36               2015   
3684    Resort Hotel           0        165               2015   
3708    Resort Hotel           0        165               2015   
...              ...         ...        ...                ...   
115029    City Hotel           0        107               2017   
115091    City Hotel           0          1               2017   
116251    City Hotel           0         44               2017   
116534    City Hotel           0          2               2017   
117087    City Hotel           0        170               2017   

       arrival_date_month  arrival_date_week_number  stays_in_weekend_nights  \
2224              October                        41                        0   
2409              October                        42                        0   
3181             November                        47                        1   
3684             December                        53                        1   
3708             December                        53                        2   
...                   ...                       ...                      ...   
115029               June                        26                        0   
115091               June                        26                        0   
116251               July                        28                        1   
116534               July                        28                        2   
117087               July                        30                        0   

        stays_in_week_nights  adults  children  babies meal country  \
2224                       3       0       0.0       0   SC     PRT   
2409                       0       0       0.0       0   SC     PRT   
3181                       2       0       0.0       0   SC     ESP   
3684                       4       0       0.0       0   SC     PRT   
3708                       4       0       0.0       0   SC     PRT   
...                      ...     ...       ...     ...  ...     ...   
115029                     3       0       0.0       0   BB     CHE   
115091                     1       0       0.0       0   SC     PRT   
116251                     1       0       0.0       0   SC     SWE   
116534                     5       0       0.0       0   SC     RUS   
117087                     2       0       0.0       0   BB     BRA   

       is_repeated_guest  previous_cancellations  \
2224                   0                       0   
2409                   0                       0   
3181                   0                       0   
3684                   0                       0   
3708                   0                       0   
...                  ...                     ...   
115029                 0                       0   
115091                 0                       0   
116251                 0                       0   
116534                 0                       0   
117087                 0                       0   

        previous_bookings_not_canceled  booking_changes deposit_type  \
2224                                 0                1   No Deposit   
2409                                 0                0   No Deposit   
3181                                 0                0   No Deposit   
3684                                 0                1   No Deposit   
3708                                 0                1   No Deposit   
...                                ...              ...          ...   
115029                               0                1   No Deposit   
115091                               0                0   No Deposit   
116251                               0                2   No Deposit   
116534                               0                1   No De

It does not make sense to have a booking with 0 guests. As 180 rows is not a lot compared to total number of observations (0,15%), these rows should be removed.

In [28]:
# Sense check for observations with zero total number of nights
sense_check_nights = df_bookings_clean2[df_bookings_clean2['total_number_nights']<1]

In [29]:
sense_check_nights

hotel is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel           0        342               2015   
1       Resort Hotel           0        737               2015   
167     Resort Hotel           0        111               2015   
168     Resort Hotel           0          0               2015   
196     Resort Hotel           0          8               2015   
...              ...         ...        ...                ...   
115483    City Hotel           0         15               2017   
117701    City Hotel           0          0               2017   
118029    City Hotel           0          0               2017   
118631    City Hotel           0         78               2017   
118963    City Hotel           0          1               2017   

       arrival_date_month  arrival_date_week_number  stays_in_weekend_nights  \
0                    July                        27                        0   
1                    July                        27                        0   
167                  July                        28                        0   
168                  July                        28                        0   
196                  July                        28                        0   
...                   ...                       ...                      ...   
115483               July                        27                        0   
117701             August                        32                        0   
118029             August                        33                        0   
118631             August                        34                        0   
118963             August                        35                        0   

        stays_in_week_nights  adults  children  babies meal country  \
0                          0       2       0.0       0   BB     PRT   
1                          0       2       0.0       0   BB     PRT   
167                        0       2       0.0       0   BB     PRT   
168                        0       1       0.0       0   BB     PRT   
196                        0       2       0.0       0   BB     PRT   
...                      ...     ...       ...     ...  ...     ...   
115483                     0       1       0.0       0   SC     FRA   
117701                     0       2       0.0       0   BB     PRT   
118029                     0       2       0.0       0   BB     PRT   
118631                     0       1       0.0       0   BB     PRT   
118963                     0       2       0.0       0   BB     PRT   

       is_repeated_guest  previous_cancellations  \
0                      0                       0   
1                      0                       0   
167                    0                       0   
168                    0                       0   
196                    0                       0   
...                  ...                     ...   
115483                 0                       0   
117701                 1                       0   
118029                 1                       0   
118631                 0                       0   
118963                 0                       0   

        previous_bookings_not_canceled  booking_changes deposit_type  \
0                                    0                3   No Deposit   
1                                    0                4   No Deposit   
167                                  0                0   No Deposit   
168                                  0                0   No Deposit   
196                                  0                0   No Deposit   
...                                ...              ...          ...   
115483                               0                0   No Deposit   
117701                               0                0   No Deposit   
118029                               0                0   No Deposit   
118631                               0                7   No De

It does not make sense to have a booking with 0 nights. As 715 rows is not a lot compared to total number of observations (0,59%), these rows should be removed.

# 6. Data Cleaning after data sense checks

In [30]:
# Removing observations with adr < 0
df_bookings_clean3 = df_bookings_clean2[(df_bookings_clean2['adr'] < 0) == False]

In [31]:
# Removing observations with total guests < 1
df_bookings_clean4 = df_bookings_clean3[(df_bookings_clean3['total_guests'] < 1) == False]

In [32]:
# Removing observations with total guests < 1
df_bookings_clean5 = df_bookings_clean4[(df_bookings_clean4['total_number_nights'] < 1) == False]

In [33]:
# Sense check after data cleaning
df_bookings_clean5.describe()

lead_time  arrival_date_year  arrival_date_week_number  \
count  118560.000000      118560.000000             118560.000000   
mean      104.508494        2016.157667                 27.157001   
std       106.915823           0.707692                 13.589513   
min         0.000000        2015.000000                  1.000000   
25%        18.000000        2016.000000                 16.000000   
50%        70.000000        2016.000000                 28.000000   
75%       161.000000        2017.000000                 38.000000   
max       709.000000        2017.000000                 53.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            118560.000000         118560.000000  118560.000000   
mean                  0.932077              2.512770       1.859995   
std                   0.995431              1.893242       0.575443   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  previous_cancellations  \
count  118560.000000  118560.000000           118560.000000   
mean        0.104200       0.007962                0.087635   
std         0.399127       0.097562                0.847190   
min         0.000000       0.000000                0.000000   
25%         0.000000       0.000000                0.000000   
50%         0.000000       0.000000                0.000000   
75%         0.000000       0.000000                0.000000   
max        10.000000      10.000000               26.000000   

       previous_bookings_not_canceled  booking_changes  days_in_waiting_list  \
count                   118560.000000    118560.000000         118560.000000   
mean                         0.137171         0.218354              2.333114   
std                          1.500721         0.637620             17.643149   
min                          0.000000         0.000000              0.000000   
25%                          0.000000         0.000000              0.000000   
50%                          0.000000         0.000000              0.000000   
75%                          0.000000         0.000000              0.000000   
max                         72.000000        18.000000            391.000000   

                 adr   total_guests  total_number_nights  
count  118560.000000  118560.000000        118560.000000  
mean      102.527187       1.972158             3.444846  
std        50.003554       0.719098             2.534904  
min         0.000000       1.000000             1.000000  
25%        70.000000       2.000000             2.000000  
50%        95.000000       2.000000             3.000000  
75%       126.000000       2.000000             4.000000  
max      5400.000000      55.000000            69.000000

In [34]:
# Removing observations with adr > 1000
df_bookings_clean6 = df_bookings_clean5[(df_bookings_clean5['adr'] > 1000) == False]

In [35]:
# Datatypes after data cleanign
df_bookings_clean6.dtypes

hotel                              object
is_canceled                        object
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
is_repeated_guest                  object
previous_cancellations              int64
previous_bookings_not_canceled      int64
booking_changes                     int64
deposit_type                       object
days_in_waiting_list                int64
customer_type                      object
adr                               float64
reservation_status                 object
total_guests                      float64
total_number_nights               

In [36]:
# Check of final shape
df_bookings_clean6.shape

(118559, 24)

# 7. Exporting cleaned dataframe

In [37]:
# Exporting dataframe
df_bookings_clean6.to_pickle(os.path.join(path,'02_Data','Prepared Data', 'hotel_bookings_cleaned.pkl'))